plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -357,15 +353,24 @@ maybeAddSubst body a s n rhs = do
     hints <- view iiHints
     let hinted = shouldInline hints a n
 
-    preUnconditional <- preInlineUnconditional rhs'
-    if preUnconditional
+    if hinted -- if we've been told specifically, then do it right away
     then extendAndDrop (Done $ dupable rhs')
     else do
-        -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-        postUnconditional <- postInlineUnconditional rhs'
-        if hinted || postUnconditional
+        -- See Note [Inlining and purity]
+        termIsPure <- checkPurity rhs'
+        preUnconditional <- preInlineUnconditional body s n termIsPure
+        if preUnconditional
         then extendAndDrop (Done $ dupable rhs')
-        else pure $ Just rhs'
+        -- See Note [Inlining and purity]
+        -- This is the case where we don't know that the number of occurrences is exactly one,
+        -- so there's no point checking if the term is immediately evaluated.
+        else if not termIsPure then pure $ Just rhs'
```

Comment:
Why did you move this out of `postInlineUnconditional`? It seems to be part of it, I don't see why we would make the two cases asymmetrical like this?

Reply:
I'm actually thinking about renaming these and separate them out even more (effectSafe, usedOnce etc). We don't follow the paper so closely - I'll add comments about their correspondence to the paper. What do you think?

Reply:
Well, if you take that to its logical conclusion you're just talking about inlining `preInlineUnconditional` and `postInlineUnconditional`. All those functions _are_ is a collection of checks to decide what to do. If we start taking some of the checks out from the functions, then their existence doesn't make much sense any more.

Reply:
I think maybe just try it and then take a look and see if you think it's actually clearer.